File: merge-intervals.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    vector<Interval> merge(vector<Interval>& intervals) {
        int n = intervals.size();
        vector<Interval> merged;
        // sort intervals according to first item
        for (int i = 0; i < n;) {
            int j = i, newEnd = intervals[i].end;
            for (j = i+1; intervals[j].start <= intervals[i].end; j++) {
                if (intervals[j].end > intervals[i].end) {
                    newEnd = max(newEnd, intervals[j].end);
                }
            }
            Interval interval(intervals[i].start, newEnd);
            merged.push_back(interval);
            i = j;
        }   
        return merged;
    }
};

File: merge-intervals.java
class Main {
    private class IntervalComparator implements Comparator<Integer> {
        public int compare(Interval a, Interval b) {
            return a.start < b.start ? -1 : a.start == b.start ? 0 : 1;   
        }
    }
    public List<Interval> merge(List<Interval> intervals) {
        // sort the intervals
        Collections.sort(intervals, new IntervalComparator());
        LinkedList<Interval> merged = new LinkedList<Interval>();
        for (int i = 0; i < intervals.length; i++) {
            if (merged.isEmpty() || merged.getLast().end < intervals[i].start) {
                merged.add(intervals[i]);   
            } else {
                merged.getLast().end = Math.max(merged.getLast().end, intervals[i].end);   
            }
        }
        return merged;
    }
}
